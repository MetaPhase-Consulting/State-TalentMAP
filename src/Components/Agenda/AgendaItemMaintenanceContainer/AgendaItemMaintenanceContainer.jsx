import { useEffect, useRef, useState } from 'react';
import FontAwesome from 'react-fontawesome';
import InteractiveElement from 'Components/InteractiveElement';
import { Tooltip } from 'react-tippy';
import { withRouter } from 'react-router';
import { filter, find, get, isNil } from 'lodash';
import MediaQuery from 'Components/MediaQuery';
import Spinner from 'Components/Spinner';
import { Link } from 'react-router-dom';
import AgendaItemResearchPane from '../AgendaItemResearchPane';
import AgendaItemMaintenancePane from '../AgendaItemMaintenancePane';
import AgendaItemTimeline from '../AgendaItemTimeline';
import { RG as RemarksGlossaryTabID } from '../AgendaItemResearchPane/AgendaItemResearchPane';

const AgendaItemMaintenanceContainer = (props) => {
  const researchPaneRef = useRef();

  const [legsContainerExpanded, setLegsContainerExpanded] = useState(false);
  const [agendaItemMaintenancePaneLoading, setAgendaItemMaintenancePaneLoading] = useState(true);
  const [agendaItemTimelineLoading, setAgendaItemTimelineLoading] = useState(true);
  const [spinner, setSpinner] = useState(true);

  const userRemarks = [];
  const [userSelections, setUserSelections] = useState(userRemarks);


  const updateSelection = (remark) => {
    const userSelection$ = [...userSelections];
    const found = find(userSelection$, { seq_num: remark.seq_num });
    if (!found) {
      userSelection$.push(remark);
      setUserSelections(userSelection$);
    } else {
      setUserSelections(filter(userSelection$, (r) => r.seq_num !== remark.seq_num));
    }
  };

  function toggleExpand() {
    setLegsContainerExpanded(!legsContainerExpanded);
  }

  const rotate = legsContainerExpanded ? 'rotate(0)' : 'rotate(-180deg)';

  const id = get(props, 'match.params.id'); // client's perdet
  const isCDO = get(props, 'isCDO');

  // need to update once further integration is done
  const employeeName = 'Employee Name Placeholder';
  const employeeHasCDO = !isNil(get(employeeName, 'person.cdo'));

  const updateResearchPaneTab = tabID => {
    researchPaneRef.current.setSelectedNav(tabID);
  };

  const openRemarksResearchTab = () => {
    setLegsContainerExpanded(false);
    updateResearchPaneTab(RemarksGlossaryTabID);
  };

  useEffect(() => {
    if (!agendaItemMaintenancePaneLoading && !agendaItemTimelineLoading) {
      setSpinner(false);
    }
  }, [agendaItemMaintenancePaneLoading, agendaItemTimelineLoading]);

  return (
    <div>
      <div className="aim-header-container">
        <div className="aim-title-container">
          <FontAwesome
            name="user-circle-o"
            size="lg"
          />
          Agenda Item Maintenance
          {isCDO && employeeHasCDO ?
            <span className="aim-title-dash">
                -
              <Link to={`/profile/public/${id}`}>
                <span className="aim-title">
                  {` ${employeeName}`}
                </span>
              </Link>
            </span>
            :
            <span>
              {` - ${employeeName}`}
            </span>
          }
        </div>
      </div>
      <MediaQuery breakpoint="screenXlgMin" widthType="max">
        {matches => (
          <div className={`ai-maintenance-container${matches ? ' stacked' : ''}`}>
            <div className={`maintenance-container-left${(legsContainerExpanded || matches) ? '-expanded' : ''}`}>
              {
                spinner &&
                  <Spinner type="left-pane" size="small" />
              }
              <AgendaItemMaintenancePane
                leftExpanded={(legsContainerExpanded || matches)}
                onAddRemarksClick={openRemarksResearchTab}
                perdet={id}
                unitedLoading={spinner}
                setParentLoadingState={setAgendaItemMaintenancePaneLoading}
                updateSelection={updateSelection}
                userSelections={userSelections}
              />
              <AgendaItemTimeline
                unitedLoading={spinner}
                setParentLoadingState={setAgendaItemTimelineLoading}
              />
            </div>
            <div className={`expand-arrow${matches ? ' hidden' : ''}`}>
              <InteractiveElement onClick={toggleExpand}>
                <Tooltip
                  title={legsContainerExpanded ? 'Expand Research' : 'Collapse Research'}
                  arrow
                >
                  <FontAwesome
                    style={{ transform: rotate, transition: 'all 0.65s linear' }}
                    name="arrow-circle-left"
                    size="lg"
                  />
                </Tooltip>
              </InteractiveElement>
            </div>
            <div className={`maintenance-container-right${(legsContainerExpanded && !matches) ? ' hidden' : ''}`}>
              <AgendaItemResearchPane
                perdet={id}
                ref={researchPaneRef}
                userSelections={userSelections}
                updateSelection={updateSelection}
              />
            </div>
          </div>
        )}
      </MediaQuery>
    </div>
  );
};

export default withRouter(AgendaItemMaintenanceContainer);
