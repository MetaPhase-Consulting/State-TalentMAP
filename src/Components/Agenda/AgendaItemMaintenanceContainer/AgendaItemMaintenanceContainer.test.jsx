import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import AgendaItemMaintenanceContainer from './AgendaItemMaintenanceContainer';

describe('AgendaItemMaintenanceContainerComponent', () => {
  it('is defined', () => {
    const wrapper = shallow(<AgendaItemMaintenanceContainer />);
    expect(wrapper).toBeDefined();
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<AgendaItemMaintenanceContainer />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
