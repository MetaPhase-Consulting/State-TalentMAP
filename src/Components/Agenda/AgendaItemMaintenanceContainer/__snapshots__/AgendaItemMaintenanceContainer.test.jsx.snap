// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgendaItemMaintenanceContainerComponent matches snapshot 1`] = `
<div
  className="agenda-item-maintenace-container"
>
  <Connect(withRouter(BackButton)) />
  <div
    className="ai-maintenance-containers"
  >
    <div
      className="maintenance-container-left"
    >
      Left Maintenance Container
      <div
        className="expand-arrow"
      >
        <InteractiveElement
          className=""
          onClick={[Function]}
          type="div"
        >
          <Tooltip
            animateFill={true}
            animation="shift"
            arrow={true}
            arrowSize="regular"
            className=""
            delay={0}
            disabled={false}
            distance={10}
            duration={375}
            followCursor={false}
            hideDelay={0}
            hideDuration={375}
            hideOnClick={true}
            html={null}
            inertia={false}
            interactive={false}
            interactiveBorder={2}
            multiple={false}
            offset={0}
            onHidden={[Function]}
            onHide={[Function]}
            onRequestClose={[Function]}
            onShow={[Function]}
            onShown={[Function]}
            popperOptions={Object {}}
            position="top"
            size="regular"
            sticky={false}
            stickyDuration={200}
            style={Object {}}
            tag="div"
            theme="dark"
            title="Expand container"
            touchHold={false}
            trigger="mouseenter focus"
            unmountHTMLWhenHide={false}
            zIndex={9999}
          >
            <FontAwesome
              name="arrow-circle-left"
              size="lg"
              style={
                Object {
                  "transform": "rotate(-180deg)",
                  "transition": "all 0.65s linear",
                }
              }
            />
          </Tooltip>
        </InteractiveElement>
      </div>
    </div>
    <div
      className="maintenance-container-right"
    >
      Right Maintenance Container
    </div>
  </div>
</div>
`;
